crm_facebook_leads/models/lead.py

```diff
@@ -264,32 +247,36 @@ def process_lead_field_data(self, lead):
                           if l.get('name') and l.get('values')])
         return lead_data
 
-    def lead_processing(self, r, form):
-        if not r.get('data'):
-            return
-        for lead in r['data']:
-            lead = self.process_lead_field_data(lead)
-            if not self.search([('facebook_lead_id', '=', lead.get('id')), '|', ('active', '=', True), ('active', '=', False)]):
-                self.lead_creation(lead, form)
-
-        # /!\ NOTE: Once finished a page let us commit that
-        try:
-            self.env.cr.commit()
-        except Exception:
-            self.env.cr.rollback()
-
-        if r.get('paging') and r['paging'].get('next'):
-            _logger.info('Fetching a new page in Form: %s' % form.name)
-            self.lead_processing(requests.get(r['paging']['next']).json(), form)
-        return
+    def lead_processing(self, response, form):
+        data = response.get('data', False)
+        while data:
+            # /!\ NOTE: Once finished a page let us commit that
+            with self.env.cr.savepoint():
```

Comment:
@moylop260 could you give us your opinion regarding changing commit by savepoint().

And whereelse you regard it needed please?

Regards

Reply:
JFYI:
https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#never-commit-the-transaction

Reply:
@GavyMG referencing the OCA, I may die on peace now.

Reply:
Lemme* explain myself because I didn't do it.

For me it is paramount that after eacht page is processed, if any Exception is raised the previous transactions is rolled back. Not the whole bunch of records already processed.

If this is fulfilled by the savepoint() we are done.

If that is not the case. Then I want to clearly violate the `https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#never-commit-the-transaction`

`And forgive me father for I have sinned`

And before making a sin:
Could you please @moylop260 / @luisg123v  / @GavyMG  help me understanding the difference between commit() and savepoint()? I am ignorant of the difference.

And yet I don't understand the OCA explanation. It still doesn't give me the certainty that I want to achieve with commit().

Thanks in advance for the explanation.

Regards.

Reply:
@hbto,

When `with cr.savepoint` is used, it automatically creates a savepoint:
- When an error occurs inside the `with` block, the transaction is rolled back to that savepoint
- When no error occurs, the transaction is **not** committed.

So, it's a safe way to save current work without needing to actually commit changes

Reply:
Note:  
Once the `with` block is left, the savepoint is released (destroyed). So if an error occurs after the block, and it's not inside another `with cr.savepoint()` block, the entire transaction will be rolled back.

Reply:
@luisg123v Thanks a lot.

Now I feel a little bit less ignorant.

This is now resolved to me.

Regards